### PR TITLE
partial history

### PR DESCRIPTION
**Pull request by yiminc**
add partial history support

tested 3 cases for partial history
1) happy case;
2) disable cache (need code change) to mimic the cache been evicted case
3) always destroy cached state after each task (need code change) to mimic the error case
All cases worked.

open issues:
1) need better eviction policy for LRU, for now using fixed size (1000)
2) query task is not sticky (need minor server side change to make sure query task does not include new events that could potentially change state)
3) client side need update once query task is possible to be sticky
4) close decision could fail, but cache is unloaded anyway, need server side change to return error/flag to indicate that.
5) need way to remove stickiness in case of error.

### CODE REVIEW

**Comment by yiminc** [comment]
tested 3 cases for partial history
1) happy case;
2) disable cache (need code change) to mimic the cache been evicted case
3) always destroy cached state after each task (need code change) to mimic the error case
All cases worked.

**Comment by mfateev** [comment]
I'm confused. The change contains refactoring of the iterator only. But the description claims that it contains the full support for workflow state caching.

**Comment by yiminc** [comment]
The changes for caching is in internal_task_handlers.go. The diff view collapse the file when there is more than a few hundred lines of changes.

File: internal_task_handlers.go [hunk 1]
```diff
@@ -308,83 +324,242 @@ func newWorkflowTaskHandler(
...
-// ProcessWorkflowTask processes each all the events of the workflow task.
-func (wth *workflowTaskHandlerImpl) ProcessWorkflowTask(
-	task *s.PollForDecisionTaskResponse,
-	getHistoryPage GetHistoryPage,
-	emitStack bool,
-) (result interface{}, stackTrace string, err error) {
-	if task == nil {
-		return nil, "", errors.New("nil workflowtask provided")
+// TODO: need a better eviction policy based on memory usage
+var workflowCache = cache.New(1000, &cache.Options{
```
Comment (mfateev) [inline internal_task_handlers.go]: Shouldn't cache be per worker?

File: internal_task_handlers.go [hunk 2]
```diff
@@ -308,83 +324,242 @@ func newWorkflowTaskHandler(
...
+	workflowContext := &workflowExecutionContext{workflowStartTime: wfStartTime, workflowInfo: workflowInfo, wth: wth}
+	workflowContext.resetWorkflowState()
 
-	isWorkflowCompleted := false
-	var completionResult []byte
-	var failure error
+	return workflowContext, nil
+}
+
+func (wth *workflowTaskHandlerImpl) getOrCreateWorkflowContext(task *s.PollForDecisionTaskResponse,
```
Comment (mfateev) [inline internal_task_handlers.go]: While I couldn't find any obvious problems with this method I believe it could be refactored to improve it readability.

Comment (yiminc) [reply internal_task_handlers.go]: the problem is the query is not sticky for now which make it more complex than it needs to be.

File: internal_task_handlers.go (hunk 1 above)
Comment (yiminc) [reply internal_task_handlers.go]: i think it should be per process, or we could have multiple caches which could make it hard if we ever want to do something based on memory stats (which is global view).

**Comment by mfateev** [comment]
OK. I don't have any more questions. Approved.

**Pull request merged**